packages/dd-trace/src/appsec/index.js

```diff
@@ -220,12 +217,46 @@ function onPassportVerify ({ credentials, user }) {
   passportTrackEvent(credentials, user, rootSpan, config.appsec.eventTracking.mode)
 }
 
+const responseAnalyzedSet = new WeakSet()
+const responseBlockedSet = new WeakSet()
+
+function onResponseWriteHead ({ req, res, abortController, statusCode, responseHeaders }) {
+  // avoid "write after end" error
+  if (responseBlockedSet.has(res)) {
+    abortController?.abort()
+    return
+  }
+
+  // avoid double waf call
+  if (responseAnalyzedSet.has(res)) {
+    return
+  }
+
+  const rootSpan = web.root(req)
+  if (!rootSpan) return
+
+  responseHeaders = Object.assign({}, responseHeaders)
+  delete responseHeaders['set-cookie']
+
+  const results = waf.run({
+    persistent: {
+      [addresses.HTTP_INCOMING_RESPONSE_CODE]: '' + statusCode,
+      [addresses.HTTP_INCOMING_RESPONSE_HEADERS]: responseHeaders
+    }
+  }, req)
+
+  responseAnalyzedSet.add(res)
+
+  handleResults(results, req, res, rootSpan, abortController)
+}
+
 function handleResults (actions, req, res, rootSpan, abortController) {
   if (!actions || !req || !res || !rootSpan || !abortController) return
 
   const blockingAction = getBlockingAction(actions)
   if (blockingAction) {
     block(req, res, rootSpan, abortController, blockingAction)
+    responseBlockedSet.add(res)
```

Comment:
If we add `res` in the set, when we are not really blocking, we could maintain the request opened forever. I think that it cannot apply for the current use case, but as this code is in the generic `handleResults` and could be used in the future for different blockings, I guess that we should protect this line.


```
    if (abortController?.signal.aborted) {
// or maybe...
    if (!abortController || abortController.signal.aborted) {
      responseBlockedSet.add(res)
    }
```

Reply:
```suggestion
    if (!abortController || abortController.signal.aborted) {
      responseBlockedSet.add(res)
    }
```

Reply:
Good point. I'd say when you move this set in the RASP PR, you can make it a bit safer by calling it directly from blocking.js